packages/material-ui-utils/src/elementTypeAcceptingRef.js

```diff
@@ -0,0 +1,50 @@
+import * as PropTypes from 'prop-types';
+import React from 'react';
+import { isLazy, isMemo } from 'react-is';
+import chainPropTypes from './chainPropTypes';
+
+function isClassComponent(elementType) {
+  // elementType.prototype?.isReactComponent
+  const { prototype = {} } = elementType;
+
+  return Boolean(prototype.isReactComponent);
+}
+
+function elementTypeAcceptingRef(props, propName, componentName, location, propFullName) {
+  const propValue = props[propName];
+  const safePropName = propFullName || propName;
+
+  if (propValue == null) {
+    return null;
+  }
+
+  let warningHint;
+
+  /**
+   * blacklisting instead of whitelisting
+   *
+   * blacklisting will miss some components like React.Fragment. Those will at least
+   * trigger a warning in react.
+   * We can't whitelist because there is no safe way to detect React.forwardRef
+   * or class components. Safe means there's no public API.
+   *
+   */
+  if (isLazy(propValue)) {
+    warningHint = 'But you passed a React.lazy component.';
+  } else if (isMemo(propValue)) {
+    warningHint = 'But you passed a React.memo component.';
+  } else if (typeof propValue === 'function' && !isClassComponent(propValue)) {
```

Comment:
Is this warning redundant with PropTypes.elementType?

Reply:
`elementType` accepts any valid element types such as function components. These would fail here because they cannot hold a ref.

Reply:
Can we have a function this is a class component too?

Reply:
Made the same mistake. For some reason I thought the class components where instantiated but what we actually pass around is only the constructor function. This is why `[PropTypes.func, PropTypes.string]` was equivalent to `PropTypes.elementType` until exotic components like `forwardRef` were introduced.

Reply:
Yeah, I have tried the pull request tests locally before saying more bullshit.